### ISSUE TITLE
Disable Lightbar

### ISSUE THREAD

**Issue by hypeFLOW** [Ryochan7/DS4Windows#855]
Is there anyway to disable the light bar. Playing the Witcher 3 which supports DS4 natively, however I just wanted to re-map the SHARE button. Yet once I do this, the light bar colour overides what it is in game.

**Comment by Ryusennin** [comment]
#831

**Comment by Ryochan7** [comment]
Have not looked into it yet but that is sort of a different issue from 831. Do some games actually provide contextual information via the lightbar color? Rocket League does not use the lightbar so there is no need to worry about it there. I don't see any reason to worry about this type of problem unless the lightbar color does actually have some meaning in a game.

**Comment by Ryusennin** [comment]
Some PC games might support Razer Chroma / Philips Hue, but I'd be surprised if any of them would be compatible with the Sony lightbar API (even through Steam).

At least Witcher 3 (GOG version) doesn't interact with the DualShock -- you need an XInput emu like DS4W to make it recognize the controller.

**Comment by Ryochan7** [comment]
People have told me that later builds of The Witcher 3 natively support the DS4 with proper PS button prompts. PCGamingWiki seems to also indicate that the game natively supports the DS4 including lightbar support. I don't own a copy of Witcher 3 so I cannot see for myself.

**Comment by hypeFLOW** [comment]
The Witcher 3 does indeed have DS4 support and it also has light bar support. For example it changes colour when you change signs in the game.

**Comment by Ryusennin** [comment]
You mean the Steam version? Because GOG has no native support for the DS4.

**Comment by hypeFLOW** [comment]
Nope GOG version. I started playing it last year again on PC and it always had DS4 prompts with my version when connected via USB.

**Comment by Ryochan7** [comment]
> The Witcher 3 does indeed have DS4 support and it also has light bar support. For example it changes colour when you change signs in the game.

Rats. I guess it is worth looking into then.

**Comment by Ryochan7** [comment]
I can put myself in line with the Linux cucks now. I used some recent donations to pick up a copy of The Witcher 3. The GOTY edition is currently on sale for $15 at GOG. This will at least give me a target to test. Never finished Witcher 1 nor 2.

https://www.reddit.com/r/linux_cucks/

**Comment by Ryusennin** [comment]
I platinumed it on PS4. Best game of the gen imo. Play it and come back in a month!

I even made a specific DS4W profile to mimic the PS4 controls (swipe up/down to call the map).

**Comment by Ryochan7** [comment]
Took a long time to download it but it is installed now. I can confirm that DS4 button glyphs show up and the game does change the lightbar color. I have only tried it in native mode via USB rather than using DS4Windows to emulate a virtual DS4.

**Comment by Ryochan7** [comment]
Can't get The Witcher 3 to use the DS4 wirelessly. It doesn't work via BT nor the Sony Wireless Adapter. Both the DS4 v.1 and v.2 work when connected via USB.

**Comment by Ryochan7** [comment]
This is another good use case for DS4 emulation. The controls while using native DS4 support seem a bit sluggish compared to having The Witcher 3 use an emulated DS4 created by DS4Windows (USB on both). Wireless support does not work for me when using native DS4 support in the game so DS4Windows has been required to play wirelessly. I have only played about 1 hour of the game so far.

I have not tested removing the DS4Windows lightbar routine to make sure The Witcher 3 could change the DS4 lightbar color from an emulated controller. The hooks are in place in DS4Windows to relay any set lightbar color so it should technically work if DS4Windows stops changing the assigned lightbar color..

**Comment by Ryochan7** [comment]
I was wrong in the previous post. The lightbar color sent from ViGEmBus is not used in the event hook. I just made the edit to that routine to test it. Luckily, ViGEmBus does send the lightbar color to DS4Windows so it can be transferred to the real controller.

It does look like rumble events and lightbar color change events conflict in ViGEmBus. I can see the lightbar color change state in the console when using Signs in game but extra rumble events send a black lightbar color code so that color gets used instead. The lightbar color will change to green after the cooldown period is over. The lightbar color change works just fine when pulling out swords though.

**Comment by Ryochan7** [comment]
ViGEmBus really needs to provide a flag to distinguish between rumble and lightbar events. I have a workaround to try to interpret which event has activated but it is likely to be wrong if an explicit black lightbar color code is sent. It works well enough for casting Signs in The Witcher 3 though. Here is an example of what I am looking at.

```
2019-11-10T21:24:18.4580262-06:00 Large: 0 Small: 0
Red: 0 Green: 255 Blue: 0
2019-11-10T21:25:50.9225971-06:00 Large: 0 Small: 0
Red: 0 Green: 0 Blue: 255
2019-11-10T21:25:50.9225971-06:00 Large: 127 Small: 0
Red: 0 Green: 0 Blue: 0
2019-11-10T21:25:50.9629695-06:00 Large: 178 Small: 0
Red: 0 Green: 0 Blue: 0
2019-11-10T21:25:51.6221575-06:00 Large: 0 Small: 0
Red: 0 Green: 0 Blue: 0
2019-11-10T21:25:53.0744993-06:00 Large: 51 Small: 0
Red: 0 Green: 0 Blue: 0
2019-11-10T21:25:53.2682703-06:00 Large: 0 Small: 0
Red: 0 Green: 0 Blue: 0
2019-11-10T21:25:53.5211297-06:00 Large: 0 Small: 0
Red: 0 Green: 255 Blue: 0
```

**Comment by StevenMartelly** [comment]
> ViGEmBus really needs to provide a flag to distinguish between rumble and lightbar events. I have a workaround to try to interpret which event has activated but it is likely to be wrong if an explicit black lightbar color code is sent. It works well enough for casting Signs in The Witcher 3 though. Here is an example of what I am looking at.
> 
> ```
> 2019-11-10T21:24:18.4580262-06:00 Large: 0 Small: 0
> Red: 0 Green: 255 Blue: 0
> 2019-11-10T21:25:50.9225971-06:00 Large: 0 Small: 0
> Red: 0 Green: 0 Blue: 255
> 2019-11-10T21:25:50.9225971-06:00 Large: 127 Small: 0
> Red: 0 Green: 0 Blue: 0
> 2019-11-10T21:25:50.9629695-06:00 Large: 178 Small: 0
> Red: 0 Green: 0 Blue: 0
> 2019-11-10T21:25:51.6221575-06:00 Large: 0 Small: 0
> Red: 0 Green: 0 Blue: 0
> 2019-11-10T21:25:53.0744993-06:00 Large: 51 Small: 0
> Red: 0 Green: 0 Blue: 0
> 2019-11-10T21:25:53.2682703-06:00 Large: 0 Small: 0
> Red: 0 Green: 0 Blue: 0
> 2019-11-10T21:25:53.5211297-06:00 Large: 0 Small: 0
> Red: 0 Green: 255 Blue: 0
> ```

So, Did you solve this problem? I started to play in Detroit Become Human and I needed to set up dead zones. I could only do this through DS4Windows but it is trying to control the lightbar because of which it starts to blink. It is very annoying, can I fix this somehow? And thanks anyway for your work :)

**Comment by mika-n** [comment]
Blinking lightbar? Have you enabled lightbar "warning" of low battery or have you enabled "flash lightbar in high latency" option?   If the lightbar is blinking beause lag then Logs tabpage of DS4Windows should have warning messages.

**Comment by StevenMartelly** [comment]
Yes, its enabled but it doesn't matter because i am connected via USB, not Bluetooth

**Comment by Ryochan7** [comment]
I still have that test diff around somewhere. I have not looked into this further since the previous post. Been working on a few side projects and getting the WPF build of DS4Windows more finalized; the WPF build has taken up most of my dev time. I have not even played any more Witcher 3 since that day.

I have also been a bit hesitant to edit any portion of the current lightbar code.The diff just skips the routine all together. That portion of the code is a mess and I'm sure I would break the routine the first time around trying to refactor it. A simple passthrough flag would probably be sufficient to have the lightbar color generation code get skipped. Although, I am not sure if it would be better to establish some kind of extra mode setting instead.

**Comment by Ryochan7** [comment]
The old test diff was lost after several Windows re-installs. I want to restart The Witcher 3 so I spent some time today working on this problem again. The current version is a bit better than what I had made before. There is now a passthru mode for the lightbar that will let other applications have control over what the lightbar color is. Refactoring the lightbar routine to make this work was a major pain.

On a semi-related note, the Normal lightbar mode has been changed to allow the value of -1 to be set for the flashAt field. That will effectively disable the flash and pulse lightbar functionality
